### ISSUE TITLE
non-lowercased repository names cause instable team memberships

### ISSUE THREAD

**Issue by skoenig** [mineiros-io/terraform-github-team#31]
Hello, I have encountered an issue with repository names which are not all lowercase. On every application of the plan the terraform-github-team module would destroy and re-create the `github_team_membership` resources, causing members of the regarding team loosing permissions for a brief period of time, but more importantly it clears the notification settings ("Participating and @mentions" vs. "All Activity").

Here is a setup to reproduce this behavior:
`main.tf`:
```
provider "github" {
  version      = "~> 3.1"
  token        = var.github_token
  organization = "my-org"
}
```
`repositories.tf`:
```
variable "repositories" {
  default = {
    snake_case  = {
      description            = "This is a snake_case named repository"
      delete_branch_on_merge = true
    }
    camelCase = {
    }
    UPPERCASE = {
      description            = "All UPPERCASED repository"
      delete_branch_on_merge = true
      has_downloads          = false
      has_issues             = false
      has_projects           = false
      has_wiki               = false
    }
  }
}

locals {
  default_repository = {
    description            = ""
    topics                 = []
    visibility             = "private"
    allow_merge_commit     = true
    allow_rebase_merge     = true
    delete_branch_on_merge = false
    has_downloads          = true
    has_issues             = true
    has_projects           = true
    has_wiki               = true
    is_template            = false
    vulnerability_alerts   = false
    archive_on_destroy     = true
    archived               = false
  }

  repositories = {
    for k, v in var.repositories :
    k => merge(local.default_repository, v)
  }
}

resource "github_repository" "repository" {
  for_each               = local.repositories
  name                   = each.key
  description            = each.value["description"]
  topics                 = each.value["topics"]
  visibility             = each.value["visibility"]
  allow_merge_commit     = each.value["allow_merge_commit"]
  allow_rebase_merge     = each.value["allow_rebase_merge"]
  delete_branch_on_merge = each.value["delete_branch_on_merge"]
  has_downloads          = each.value["has_downloads"]
  has_issues             = each.value["has_issues"]
  has_projects           = each.value["has_projects"]
  has_wiki               = each.value["has_wiki"]
  is_template            = each.value["is_template"]
  vulnerability_alerts   = each.value["vulnerability_alerts"]
  archive_on_destroy     = each.value["archive_on_destroy"]
  archived               = each.value["archived"]
}
```
`team.tf`:
```
module "team" {
  source  = "mineiros-io/team/github"
  version = "0.2.0"
  privacy = "closed"

  name = "Awesome Team"
  members = [
    "skoenig",
  ]
  push_repositories = [
    github_repository.repository["snake_case"].name,
    github_repository.repository["camelCase"].name,
    github_repository.repository["UPPERCASE"].name,
  ]
}
```
Applying this manifests would produce something like this example output showing that 'UPPERCASE' and 'camelCase' repos are affected but not the all lowercase 'snake_case' repo:
```
module.team.github_team.team: Refreshing state... [id=4755619]
github_repository.repository["UPPERCASE"]: Refreshing state... [id=UPPERCASE]
github_repository.repository["camelCase"]: Refreshing state... [id=camelCase]
github_repository.repository["snake_case"]: Refreshing state... [id=snake_case]
module.team.github_team_membership.team_membership["skoenig"]: Refreshing state... [id=4755619:skoenig]
module.team.github_team_repository.team_repository["camelcase"]: Refreshing state... [id=4755619:camelcase]
module.team.github_team_repository.team_repository["uppercase"]: Refreshing state... [id=4755619:uppercase]
module.team.github_team_repository.team_repository["snake_case"]: Refreshing state... [id=4755619:snake_case]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # module.team.github_team_repository.team_repository["camelcase"] must be replaced
-/+ resource "github_team_repository" "team_repository" {
      ~ etag       = "W/\"dbe49fcf26117dea7ccef090bd19253215eb8e968cd258f22c8d84375ccd2aaa\"" -> (known after apply)
      ~ id         = "4755619:camelcase" -> (known after apply)
        permission = "push"
      ~ repository = "camelCase" -> "camelcase" # forces replacement
        team_id    = "4755619"
    }

  # module.team.github_team_repository.team_repository["uppercase"] must be replaced
-/+ resource "github_team_repository" "team_repository" {
      ~ etag       = "W/\"6e4c45495fa37323e6acfed0c1de35e2ec627659ebfef5abaa4c154f731170bf\"" -> (known after apply)
      ~ id         = "4755619:uppercase" -> (known after apply)
        permission = "push"
      ~ repository = "UPPERCASE" -> "uppercase" # forces replacement
        team_id    = "4755619"
    }

Plan: 2 to add, 0 to change, 2 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.team.github_team_repository.team_repository["uppercase"]: Destroying... [id=4755619:uppercase]
module.team.github_team_repository.team_repository["camelcase"]: Destroying... [id=4755619:camelcase]
module.team.github_team_repository.team_repository["uppercase"]: Destruction complete after 0s
module.team.github_team_repository.team_repository["uppercase"]: Creating...
module.team.github_team_repository.team_repository["camelcase"]: Destruction complete after 2s
module.team.github_team_repository.team_repository["camelcase"]: Creating...
module.team.github_team_repository.team_repository["camelcase"]: Creation complete after 4s [id=4755619:camelcase]
module.team.github_team_repository.team_repository["uppercase"]: Creation complete after 6s [id=4755619:uppercase]

Apply complete! Resources: 2 added, 0 changed, 2 destroyed.
```
Seems like the issue comes from `lower()` during the object creation in https://github.com/mineiros-io/terraform-github-team/blob/master/main.tf#L39-L41. It would be really helpful to fix this behavior as it causes very unpleasant issues when managing my organization.

**Comment by soerenmartius** [comment]
Hey @skoenig thanks for reporting this issue! Will look into it with @mariux at latest tomorrow and come back to you asap. Have a nice evening

**Comment by mariux** [comment]
@skoenig thanks for taking the time to create such an awesome bug report and also identifying the source of issue.

I can perfectly reproduce the issue in an improved test run. https://github.com/mineiros-io/terraform-github-team/pull/32/checks?check_run_id=2461480940 Will prepare a fix right away and keep the test in place for future reference.

**Comment by mariux** [comment]
fixed in #32 after a proper review and merging it, we will release `0.3.1`. The new test for idem-potency is now working. Thanks again for taking the time to report this.

**Comment by mariux** [comment]
released 0.3.1 just now ;)

so please use version constraint `version = "~> 0.3.1"` to ensure to get the latest 3.x version or `version = "0.3.1"` to get the exact fix.

we hope this fixed your issue. please report back if this is not the case.
